sonic_installer/main.py

```diff
@@ -317,5 +317,68 @@ def cleanup():
     if image_removed == 0:
         click.echo("No image(s) to remove")
 
+# Upgrade docker image
+@cli.command()
+@click.option('-y', '--yes', is_flag=True, callback=abort_if_false,
+        expose_value=False, prompt='New docker image will be installed, continue?')
+@click.option('--cleanup_image', is_flag=True, help="Clean up old docker images")
+@click.argument('container_name', metavar='<container_name>', required=True, type=click.Choice(["swss", "snmp", "lldp"]))
+@click.argument('tag', metavar='<tag>', required=True, type=click.STRING)
+@click.argument('url')
+def upgrade_docker(cleanup_image, container_name, tag, url):
+    """ Upgrade docker image from local binary or URL"""
+
+    # example images: docker-lldp-sv2:latest
+    cmd = "docker inspect --format '{{.Config.Image}}' " + container_name
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+    images = proc.stdout.read().rstrip()
+
+    # example image_name: docker-lldp-sv2
+    cmd = "echo " + images + " | cut -d ':' -f 1"
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+    image_name = proc.stdout.read().rstrip()
+
+    # example image_id: 7ed919240fb9
+    cmd = "docker images --format '{{.ID}}' " + image_name
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+    image_id = proc.stdout.read()
+
+
+    DEFAULT_IMAGE_PATH = os.path.join("/tmp/", image_name)
+
+    if url.startswith('http://') or url.startswith('https://'):
+        click.echo('Downloading image...')
+        urllib.urlretrieve(url, DEFAULT_IMAGE_PATH, reporthook)
+        image_path = DEFAULT_IMAGE_PATH
+    else:
+        image_path = os.path.join("./", url)
+
+    # make sure orchagent is in clean state if swss is to be upgraded
+    if container_name == "swss":
+        cmd = "docker exec -it " + container_name + " orchagent_restart_check"
+        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+        result = proc.stdout.read().rstrip()
+        if result != "RESTARTCHECK succeeded":
```

Comment:
It could be changed to check return code.  What is the benefit of doing that?  
The concern here is that, theoretically, the success of running orchagent_restart_check program itself doesn't indicate the warm restart readiness of orchagent.

Reply:
The benefit is if any result string from "orchagent_restart_check" changed for some reason, we don't have to change code here. 
orchagent_restart_check should return EXIT_SUCCESS code if check is successful and return EXIT_FAILURE if check is failed or some Error code if the program is not running due to other reasons.

Reply:
Both the result string and exit code are from orchagent_restart_check, but exit code may also indicate status of  orchagent_restart_check itself.   The result strings are designed for the result check purposely.  Anyway,  don't want to spend too much time debating this kind of issues.  Either one works for now, will change to use exit code to settle the issue.